docs: reformat docstrings to md

ccmao1130 opened the pull request:
## Changes Made

Reformatting docstrings from rst to md format.

## Related Issues

https://github.com/Eventual-Inc/Daft/pull/3980

## Checklist

- [ ] Documented in API Docs (if applicable)
- [ ] Documented in User Guide (if applicable)
- [ ] If adding a new documentation page, doc is added to `docs/mkdocs.yml` navigation
- [ ] Documentation builds and is formatted properly (tag @/ccmao1130 for docs review)

universalmind303 commented on daft/catalog/__init__.py:
````diff
@@ -83,9 +83,14 @@ def unregister_catalog(catalog_name: str | None) -> bool:
     Example:
-        >>> import daft
-        >>> daft.unregister_catalog("my_catalog")
+        ``` py linenums="1"
+        import daft
+
+        daft.unregister_catalog("my_catalog")
+        ```
+        ```
         True
+        ```
````
so these weren't RST, but were standard [doctest](https://docs.python.org/3/library/doctest.html) style docstrings. 

Without the `>>>` syntax, they won't run via `make doctests` anymore.

universalmind303 commented on daft/catalog/__init__.py:
```diff
@@ -72,7 +72,7 @@
     """Unregisters a catalog from the Daft catalog system.
 
-    DEPRECATED: This is deprecated and will be removed in daft >= 0.5.0; please use `daft.detach_catalog`.
+    !!! warning "DEPRECATED: This is deprecated and will be removed in daft >= 0.5.0; please use [daft.detach_catalog][daft.Session.detach_catalog]."
```
these `[daft.detach_catalog][daft.Session.detach_catalog]` don't resolve properly in any IDE. 

<img width="692" alt="image" src="https://github.com/user-attachments/assets/e53c5c73-090d-41f8-b74a-701f9c5f82ab" />

<img width="331" alt="image" src="https://github.com/user-attachments/assets/c2ec5d3e-6434-4583-818d-ac4d051e7893" />

universalmind303 replied:
I dont know of any IDE's or editors that support md admonitions or relative links in the docstrings. 

I'm definitely biased towards having a first class experience for docstrings in the IDE over the api reference. 

The admonotions are probably fine, but the relative links don't render very nicely


### zed
<img width="362" alt="image" src="https://github.com/user-attachments/assets/e52d2b8e-c242-4411-9875-084ba1586bb4" />

### vscode
<img width="679" alt="image" src="https://github.com/user-attachments/assets/1b0b9a10-63b4-4baf-a387-8d016043fc06" />

### helix
<img width="752" alt="image" src="https://github.com/user-attachments/assets/3c562215-fb55-4b9a-9ce7-8074aa40f3b4" />

ccmao1130 commented:
@universalmind303 
* updated `Example` to `Examples`
* removed `!!!` admonitions
* updated (most) references to direct links - let me know if i missed any
* `dataframe.py` docstrings were formatted in https://github.com/Eventual-Inc/Daft/pull/3980 (didn't want to do a major reformat again haha)

universalmind303 commented on daft/expressions/expressions.py:
```diff
@@ -667,11 +673,6 @@ def alias(self, name: builtins.str) -> Expression:
     def cast(self, dtype: DataTypeLike) -> Expression:
```
we should convert the conversion table here to markdown for better rendering. 

```md
        | Target →          | Null | Boolean | Integers | Floats | Decimal128 | String | Binary | Fixed-size Binary | Image | Fixed-shape Image | Embedding | Tensor | Fixed-shape Tensor | Python | List | Fixed-size List | Struct | Map | Timestamp | Date | Time | Duration |
        |-------------------|------|---------|----------|--------|------------|--------|--------|-------------------|-------|-------------------|-----------|--------|--------------------|--------|------|-----------------|--------|-----|-----------|------|------|----------|
        | **Source ↓**      |      |         |          |        |            |        |        |                   |       |                   |           |        |                    |        |      |                 |        |     |           |      |      |          |
        | Null              | Y    | Y       | Y        | Y      | Y          | Y      | Y      | Y                 | N     | N                 | Y         | N      | N                  | Y      | Y    | Y               | Y      | Y   | Y         | Y    | Y    | Y        |
        | Boolean           | Y    | Y       | Y        | Y      | N          | Y      | Y      | N                 | N     | N                 | N         | N      | N                  | Y      | N    | N               | N      | N   | N         | N    | N    | N        |
        | Integers          | Y    | Y       | Y        | Y      | Y          | Y      | Y      | N                 | N     | N                 | N         | N      | N                  | Y      | N    | N               | N      | N   | Y         | Y    | Y    | Y        |
        | Floats            | Y    | Y       | Y        | Y      | Y          | Y      | Y      | N                 | N     | N                 | N         | N      | N                  | Y      | N    | M               | N      | N   | Y         | Y    | Y    | Y        |
        | Decimal128        | Y    | N       | Y        | Y      | Y          | N      | N      | N                 | N     | N                 | N         | N      | N                  | Y      | N    | N               | N      | N   | N         | N    | N    | N        |
        | String            | Y    | N       | Y        | Y      | N          | Y      | Y      | N                 | N     | N                 | N         | N      | N                  | Y      | N    | N               | N      | N   | Y         | Y    | N    | N        |
        | Binary            | Y    | N       | Y        | Y      | N          | Y      | Y      | Y                 | N     | N                 | N         | N      | N                  | Y      | N    | N               | N      | N   | N         | N    | N    | N        |
        | Fixed-size Binary | Y    | N       | N        | N      | N          | N      | Y      | N                 | N     | N                 | N         | N      | N                  | Y      | N    | N               | N      | N   | N         | N    | N    | N        |
        | Image             | N    | N       | N        | N      | N          | N      | N      | N                 | Y     | Y                 | N         | Y      | Y                  | Y      | N    | N               | Y      | N   | N         | N    | N    | N        |
        | Fixed-size Image  | N    | N       | N        | N      | N          | N      | N      | N                 | Y     | Y                 | N         | Y      | Y                  | Y      | Y    | Y               | N      | N   | N         | N    | N    | N        |
        | Embedding         | Y    | N       | N        | N      | N          | n      | N      | N                 | N     | Y                 | N         | Y      | Y                  | Y      | Y    | Y               | N      | N   | N         | N    | N    | N        |
        | Tensor            | Y    | N       | N        | N      | N          | N      | N      | N                 | Y     | Y                 | N         | Y      | Y                  | Y      | N    | N               | Y      | N   | N         | N    | N    | N        |
        | Fixed-shape Tensor| N    | N       | N        | N      | N          | N      | N      | N                 | N     | Y                 | N         | Y      | Y                  | Y      | Y    | Y               | N      | N   | N         | N    | N    | N        |
        | Python            | Y    | Y       | Y        | Y      | N          | Y      | Y      | Y                 | Y     | Y                 | Y         | Y      | Y                  | Y      | Y    | Y               | Y      | N   | N         | N    | N    | N        |
        | List              | N    | N       | N        | N      | N          | N      | N      | N                 | N     | N                 | Y         | N      | N                  | N      | Y    | Y               | N      | Y   | N         | N    | N    | N        |
        | Fixed-size List   | N    | N       | N        | N      | N          | N      | N      | N                 | N     | Y                 | N         | N      | Y                  | N      | Y    | Y               | N      | N   | N         | N    | N    | N        |
        | Struct            | N    | N       | N        | N      | N          | N      | N      | N                 | Y     | N                 | N         | Y      | N                  | N      | N    | N               | Y      | N   | N         | N    | N    | N        |
        | Map               | N    | N       | N        | N      | N          | N      | N      | N                 | N     | N                 | Y         | N      | N                  | N      | Y    | Y               | N      | Y   | N         | N    | N    | N        |
        | Timestamp         | Y    | N       | Y        | Y      | N          | Y      | N      | N                 | N     | N                 | N         | N      | N                  | Y      | N    | N               | N      | N   | Y         | Y    | Y    | N        |
        | Date              | Y    | N       | Y        | Y      | N          | Y      | N      | N                 | N     | N                 | N         | N      | N                  | Y      | N    | N               | N      | N   | Y         | Y    | N    | N        |
        | Time              | Y    | N       | Y        | Y      | N          | Y      | N      | N                 | N     | N                 | N         | N      | N                  | Y      | N    | N               | N      | N   | N         | N    | Y    | N        |
        | Duration          | Y    | N       | Y        | Y      | N          | N      | N      | N                 | N     | N                 | N         | N      | N                  | Y      | N    | N               | N      | N   | N         | N    | N    | N        |
```

universalmind303 commented:
also one potential followup: the font we use doesn't do a great job at rendering tables inside the code blocks. If it's easy, we should switch the codeblock font to something that renders nicer

<img width="797" alt="Screenshot 2025-04-17 at 11 41 09 AM" src="https://github.com/user-attachments/assets/47e2e1ad-ce44-4d39-857f-753418e36639" />
